fdbkubernetesmonitor/monitor.go

```diff
@@ -0,0 +1,369 @@
+	// ConfigFile defines the path to the config file to load.
+	ConfigFile string
+
+	// CustomEnvironment defines the custom environment variables to use when
+	// interpreting the monitor configuration.
+	CustomEnvironment map[string]string
+
+	// ActiveConfiguration defines the active process configuration.
+	ActiveConfiguration *ProcessConfiguration
+
+	// ActiveConfigurationBytes defines the source data for the active process
+	// configuration.
+	ActiveConfigurationBytes []byte
+
+	// LastConfigurationTime is the last time we successfully reloaded the
+	// configuration file.
+	LastConfigurationTime time.Time
+
+	// ProcessIDs stores the PIDs of the processes that are running. A PID of
+	// zero will indicate that a process does not have a run loop. A PID of -1
+	// will indicate that a process has a run loop but is not currently running
+	// the subprocess.
+	ProcessIDs []int
+
+	// Mutex defines a mutex around working with configuration.
+	// This is used to synchronize access to local state like the active
+	// configuration and the process IDs from multiple goroutines.
+	Mutex sync.Mutex
+
+	// PodClient is a client for posting updates about this pod to
+	// Kubernetes.
+	PodClient *PodClient
+
+	// Logger is the logger instance for this monitor.
+	Logger logr.Logger
+}
+
+// StartMonitor starts the monitor loop.
+func StartMonitor(logger logr.Logger, configFile string, customEnvironment map[string]string) {
+	podClient, err := CreatePodClient()
+	if err != nil {
+		panic(err)
+	}
+
+	monitor := &Monitor{
+		ConfigFile:        configFile,
+		PodClient:         podClient,
+		Logger:            logger,
+		CustomEnvironment: customEnvironment,
+	}
+
+	go func() { monitor.WatchPodTimestamps() }()
+	monitor.Run()
+}
+
+// LoadConfiguration loads the latest configuration from the config file.
+func (monitor *Monitor) LoadConfiguration() {
+	file, err := os.Open(monitor.ConfigFile)
+	if err != nil {
+		monitor.Logger.Error(err, "Error reading monitor config file", "monitorConfigPath", monitor.ConfigFile)
+		return
+	}
+	defer file.Close()
+	configuration := &ProcessConfiguration{}
+	configurationBytes, err := io.ReadAll(file)
+	if err != nil {
+		monitor.Logger.Error(err, "Error reading monitor configuration", "monitorConfigPath", monitor.ConfigFile)
+	}
+	err = json.Unmarshal(configurationBytes, configuration)
+	if err != nil {
+		monitor.Logger.Error(err, "Error parsing monitor configuration", "rawConfiguration", string(configurationBytes))
+		return
+	}
+
+	if currentContainerVersion == configuration.Version {
+		configuration.BinaryPath = fdbserverPath
+	} else {
+		configuration.BinaryPath = path.Join(sharedBinaryDir, configuration.Version, "fdbserver")
+	}
+
+	binaryStat, err := os.Stat(configuration.BinaryPath)
+	if err != nil {
+		monitor.Logger.Error(err, "Error checking binary path for latest configuration", "configuration", configuration, "binaryPath", configuration.BinaryPath)
+		return
+	}
+	if binaryStat.Mode()&0o100 == 0 {
```

Comment:
I think it's worth to move this check into it's dedicated method with a clear name `isExecutableByOwner(...)`